### PR TITLE
fix 'fields and values' not accessible in expression editor in default values

### PR DESCRIPTION
**Pull request by 3nids**
even if the expression generator is set, the expression dialog needs a pointer to the vector layer to behave es expected

### CODE REVIEW

**Comment by 3nids** [comment]
the other idea would be to get the layer from the context if the layer is null in the dialog constructor
see #35031

**Comment by Gustry** [comment]
@3nids Very similar issue, I noticed in master `Field and values` is not accessible too in the visibility by expression for a container.
It's working well on QGIS 3.10.

**Comment by m-kuhn** [comment]
It was left out on purpose. Field values are not available when calculating default values.

Should this be done only when "Refresh on Update" is checked? Or are there other possibilities when using values from the current form does make sense?

**Comment by roya0045** [comment]
See #32888

**Comment by 3nids** [comment]
there is now a warning if an expression uses fields on insert:

Screen-Recording-2020-03-12-at-1

**Review by signedav — APPROVED** [review]

**Comment by 3nids** [comment]
since the gif is a bit too fast, here is the text:

> The expression is using fields on insert. This behavior is not guaranteed.

**Comment by nyalldawson** [comment]
What does "is not guaranteed" mean? It should say "will be totally broken"!

**Comment by m-kuhn** [comment]
I would say "Using fields in a default value expression only works if apply default values on update is checked"

This leaves open if someone finds a scenario in which it works for him for whatever reason and also avoids using the term "broken" in our beloved application.

**Comment by nyalldawson** [comment]
@m-kuhn

> This leaves open if someone finds a scenario in which it works for him 

That situation is impossible! There's no feature set at expression evaluation time for this, so it will NEVER work...

**Comment by nyalldawson** [comment]
Being more constructive, maybe we should only show the fields in the dialog when "on update" is checked. That at least will make it less likely for user confusion.

And then we add the strongly worded warning just in case a user changes uses "on update", creates an expression with fields, then unchecks "on update"..

**Comment by m-kuhn** [comment]
I didn't revise how it works nowadays.
Some scenarios I could imagine that it works (without having groundtruthed it):

- There is a possibility to provide a feature to a new attribute form. Maybe now or in the future attributes specified in there can be taken into account.
- Maybe the order of initialization of the default value matters (which is very close to UB)

-----------

The proposal to only show it if the checkbox has been activated has been rejected in internal discussions at the laax hackfest because

- The intuitive order to fill the widgets is from top to bottom, meaning the "update" checkbox will naturally be updated after filling the expression and there is nothing that will tell a user he should do otherwise
- This makes it look like a system with random behavior which normally does not increase my trust in an application

**Comment by nyalldawson** [comment]
> Some scenarios I could imagine that it works (without having groundtruthed it):

But neither of these reflect current reality. In current master, you can NEVER use field values as part of non-on-update defaults.

> This makes it look like a system with random behavior which normally does not increase my trust in an application

I agree with this. And my concern is that just saying "not guaranteed" we are lying to users and going to cause a whole lot of wasted time and frustration. A user will likely keep smashing away at their expression thinking THEY are doing something wrong, when they'll actually never be able to fix their expression...

**Comment by nyalldawson** [comment]
To take an extreme example, I think this message is just as destructive as it would be if we added a button on the toolbar which did nothing but show a message box "this button action is not guaranteed. Keep pressing it, maybe it'll work sometime!"

**Comment by roya0045** [comment]
I agree with Nyall on this, usually I'm in favor of letting the users do more and not baby-proofing thing but this is taking things too far as most use case will result in confusion due to a seemingly inconsistent behavior.

**Comment by m-kuhn** [comment]
Now I am a bit confused. I agree with both of you that we shouldn't use "not guaranteed" and we shouldn't have a confusing behavior. Can we still try to find something that does not involve the term "broken" as I tried to come up with in https://github.com/qgis/QGIS/pull/35030#issuecomment-598343631?

File: src/ui/attributeformconfig/qgsattributetypeedit.ui [hunk 1]
```diff
@@ -184,36 +184,65 @@
...
+           </property>
+           <property name="pixmap">
+            <pixmap resource="../../../images/images.qrc">:/images/themes/default/mIconWarning.svg</pixmap>
+           </property>
+          </widget>
+         </item>
+         <item>
+          <widget class="QLabel" name="label_7">
+           <property name="text">
+            <string>The expression is using fields on insert. This behavior is not guaranteed.</string>
```
Comment (3nids) [inline src/ui/attributeformconfig/qgsattributetypeedit.ui]: following @m-kuhn's proposition
```suggestion
            <string>Using fields in a default value expression only works if apply default values on update is checked.</string>
```

Comment (nyalldawson) [reply src/ui/attributeformconfig/qgsattributetypeedit.ui]: That's looking better:

"Using fields in a default value expression only works if "Apply default value on update" is checked."

Comment (3nids) [reply src/ui/attributeformconfig/qgsattributetypeedit.ui]: thanks

**Pull request merged**